Lib/functools.py

```diff
@@ -817,6 +817,10 @@ def register(cls, func=None):
         return func
 
     def wrapper(*args, **kw):
+        if not args:
+            raise TypeError(f'{func.__name__} requires at least '
```

Comment:
`func` can be not having the `__name__` attribute in general case. This PR can break a code which use `singledispatch()` with custom callables.

Reply:
Also this change creates a new reference to `func` linked to from `wrapper`. It may be better to keep a reference just to the name. E.g.
```python
funcname = getattr(func, '__name__', 'singledispatch function')
def wrapper(*args, **kw):
    # use funcname
```